File: ios/WaveformVisualization.swift
import AVFoundation

class WaveformVisualization: AudioVisualization {
    private let waveformLayer = CAShapeLayer()

    var layer: CALayer {
        return waveformLayer
    }
    
    init() {
        setupWaveformLayer()
    }
    
    private func setupWaveformLayer() {
        waveformLayer.strokeColor = UIColor.blue.cgColor
        waveformLayer.fillColor = nil
        waveformLayer.lineWidth = 2.0
    }
    
    func getSamplesFromAudio(_ buffer: AVAudioPCMBuffer) -> [[Float]] {
        guard let channelData = buffer.floatChannelData else {
            print("Error: Could not access float channel data")
            return []
        }

        let sampleCount = Int(waveformLayer.bounds.width / waveformLayer.lineWidth)
        let channelCount = Int(buffer.format.channelCount)
        let frameLength = Int(buffer.frameLength)

        // Calculate how many full pieces we can create
        let pieceCount = frameLength / sampleCount

        var samplePieces: [[Float]] = []

        for pieceIndex in 0..<pieceCount {
            var pieceSamples: [Float] = []

            let startIndex = pieceIndex * sampleCount
            let endIndex = min(startIndex + sampleCount, frameLength)

            for i in startIndex..<endIndex {
                var sample: Float = 0
                for channel in 0..<channelCount {
                    sample += abs(channelData[channel][i])
                }
                sample /= Float(channelCount)
                pieceSamples.append(sample)
            }

            // Adjust sample count if necessary
            while pieceSamples.count > sampleCount {
                pieceSamples.removeLast()
            }
            while pieceSamples.count < sampleCount {
                pieceSamples.append(0)
            }

            samplePieces.append(pieceSamples)
        }

        // Handle any remaining samples
        if frameLength % sampleCount != 0 {
            var remainingSamples: [Float] = []

            let startIndex = pieceCount * sampleCount

            for i in startIndex..<frameLength {
                var sample: Float = 0
                for channel in 0..<channelCount {
                    sample += abs(channelData[channel][i])
                }
                sample /= Float(channelCount)
                remainingSamples.append(sample)
            }

            // Adjust sample count if necessary
            while remainingSamples.count > sampleCount {
                remainingSamples.removeLast()
            }
            while remainingSamples.count < sampleCount {
                remainingSamples.append(0)
            }

            samplePieces.append(remainingSamples)
        }

        return samplePieces
    }
    
    func updateVisualization(with samples: [Float]) {
        let path = UIBezierPath()
        let width = waveformLayer.bounds.width
        let height = waveformLayer.bounds.height
        let midPoint = height / 2
        let sampleWidth = width / CGFloat(samples.count)
        
        for (index, sample) in samples.enumerated() {
            let x = CGFloat(index) * sampleWidth
            let sampleHeight = min(CGFloat(sample) * height * 1.5, height)
            
            path.move(to: CGPoint(x: x, y: midPoint - sampleHeight/2))
            path.addLine(to: CGPoint(x: x, y: midPoint + sampleHeight/2))
        }
        waveformLayer.path = path.cgPath
    }
    
    func clearVisualization() {
        waveformLayer.path = nil
    }
    
    func setColor(_ color: UIColor) {
        waveformLayer.strokeColor = color.cgColor
    }
    
    func setFrame(_ frame: CGRect) {
        waveformLayer.frame = frame
    }
}

File: android/src/main/java/com/paulingalls/realtimeaudio/WaveformVisualization.kt
package com.paulingalls.realtimeaudio

import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Path
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

class WaveformVisualization : AudioVisualization {
    private val waveformPaint: Paint = Paint()
    private val waveformPath: Path = Path()
    private var waveformData: FloatArray = FloatArray(0)

    init {
        waveformPaint.apply {
            color = Color.BLUE
            style = Paint.Style.STROKE
            strokeWidth = 2f
            isAntiAlias = true
        }
    }

    override fun getSamplesFromAudio(buffer: FloatArray, channelCount: Int, sampleCount: Int): ArrayList<FloatArray> {
        val frameLength = buffer.size / channelCount
        val pieceCount = frameLength / sampleCount
        val samplePieces = ArrayList<FloatArray>()

        for (pieceIndex in 0 until pieceCount) {
            val pieceSamples = mutableListOf<Float>()

            val startIndex = pieceIndex * sampleCount
            val endIndex = min(startIndex + sampleCount, frameLength)

            for (i in startIndex until endIndex) {
                var sample = 0f
                for (channel in 0 until channelCount) {
                    sample += buffer[i * channelCount + channel]
                }
                sample /= channelCount.toFloat()
                pieceSamples.add(sample)
            }
            samplePieces.add(pieceSamples.toFloatArray())
        }

        if (frameLength % sampleCount != 0) {
            val remainingSamples = mutableListOf<Float>()

            val startIndex = pieceCount * sampleCount

            for (i in startIndex until frameLength) {
                var sample = 0f
                for (channel in 0 until channelCount) {
                    sample += buffer[i * channelCount + channel]
                }
                sample /= channelCount.toFloat()
                remainingSamples.add(sample)
            }
            samplePieces.add(remainingSamples.toFloatArray())
        }

        return samplePieces
    }


    override fun updateData(data: FloatArray) {
        waveformData = data
    }

    override fun draw(canvas: Canvas, width: Float, height: Float) {
        if (waveformData.isEmpty()) return

        val centerY = height / 2
        waveformPath.reset()
        val step = max(1, waveformData.size / width.toInt())
        var x = 0f

        waveformPath.moveTo(0f, centerY)

        for (i in waveformData.indices step step) {
            val sampleValue = waveformData[i]
            val y = centerY + (sampleValue * height / 2)
            waveformPath.lineTo(x, y)
            x += width / (waveformData.size / step)
        }

        canvas.drawPath(waveformPath, waveformPaint)
    }

    override fun setColor(color: Int) {
        waveformPaint.color = color
    }
}